nifi-extension-bundles/nifi-azure-bundle/nifi-azure-registry-clients/src/main/java/org/apache/nifi/azure/devops/AzureDevOpsRepositoryClient.java

```diff
@@ -0,0 +1,739 @@
+    private static final String JSON_FIELD_COMMIT_ID = "commitId";
+    private static final String JSON_FIELD_OBJECT_ID = "objectId";
+
+    // Change request constants
+    private static final String CHANGE_TYPE_ADD = "add";
+    private static final String CHANGE_TYPE_EDIT = "edit";
+    private static final String CHANGE_TYPE_DELETE = "delete";
+    private static final String CONTENT_TYPE_BASE64 = "base64encoded";
+
+    // Common query parameter names and values
+    private static final String VERSION_DESCRIPTOR_VERSION = "versionDescriptor.version";
+    private static final String VERSION_DESCRIPTOR_VERSION_TYPE = "versionDescriptor.versionType";
+    private static final String VERSION_TYPE_BRANCH = "branch";
+    private static final String RECURSION_LEVEL = "recursionLevel";
+    private static final String RECURSION_LEVEL_ONE_LEVEL = "OneLevel";
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private final String organization;
+    private final String project;
+    private final String repoName;
+    private final URL apiUrl;
+    private final String repoPath;
+    private final String clientId;
+    private final OAuth2AccessTokenProvider tokenProvider;
+    private WebClientServiceProvider webClient;
+
+    private final boolean canRead;
+    private final boolean canWrite;
+
+    /**
+     * Constructs an AzureDevOpsRepositoryClient with the provided configuration.
+     *
+     * @param builder The builder containing the configuration for the client.
+     * @throws FlowRegistryException if any required parameters are missing or if
+     *                               access validation fails.
+     */
+    private AzureDevOpsRepositoryClient(final Builder builder) throws FlowRegistryException {
+        apiUrl = Objects.requireNonNull(builder.apiUrl, "API URL required");
+        organization = Objects.requireNonNull(builder.organization, "Organization required");
+        project = Objects.requireNonNull(builder.project, "Project required");
+        repoName = Objects.requireNonNull(builder.repoName, "Repository Name required");
+        repoPath = builder.repoPath;
+        tokenProvider = Objects.requireNonNull(builder.oauthService, "OAuth2 Access Token Provider required");
+
+        clientId = Objects.requireNonNull(builder.clientId, "Client ID is required");
+        webClient = Objects.requireNonNull(builder.webClient, "Web Client is required");
+
+        // attempt to retrieve repository information to validate access
+        final URI uri = getUriBuilder().build();
+        final JsonNode response = executeGet(uri);
+        final String projectId = response.get(JSON_FIELD_PROJECT).get(JSON_FIELD_ID).asText();
+        final String repoId = response.get(JSON_FIELD_ID).asText();
+
+        canRead = true;
+        canWrite = hasWriteAccess(projectId, repoId);
+
+        LOGGER.info("Created {} for clientId = [{}], repository [{}]", getClass().getSimpleName(), clientId, repoName);
```

Comment:
The `clientId` seems like it could be confused with the Client ID for authentication. Also if it is only used for logging, is it needed?